core/tee/tee_fs_rpc.c

```diff
@@ -0,0 +1,605 @@
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee/tee_fs.h>
+#include <tee/tee_fs_rpc.h>
+#include <trace.h>
+#include <util.h>
+
+#define RPC_FAILED -1
+
+struct tee_fs_dir {
+	int nw_dir;
+	struct tee_fs_dirent d;
+};
+
+static TEE_Result tee_fs_rpc_send_cmd(int cmd_id, struct tee_fs_rpc *bf_cmd,
+				      void *data, uint32_t len, uint32_t mode)
+{
+	TEE_Result ret;
+	struct optee_msg_param params;
+	paddr_t phpayload = 0;
+	uint64_t cpayload = 0;
+	struct tee_fs_rpc *bf;
+	int res = TEE_ERROR_GENERIC;
+
+	assert(cmd_id == OPTEE_MSG_RPC_CMD_FS ||
+	       cmd_id == OPTEE_MSG_RPC_CMD_SQL_FS);
+
+	thread_rpc_alloc_payload(sizeof(struct tee_fs_rpc) + len,
+				 &phpayload, &cpayload);
+	if (!phpayload)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!ALIGNMENT_IS_OK(phpayload, struct tee_fs_rpc))
+		goto exit;
+
+	bf = phys_to_virt(phpayload, MEM_AREA_NSEC_SHM);
+	if (!bf)
+		goto exit;
+
+	memset(&params, 0, sizeof(params));
+	params.attr = OPTEE_MSG_ATTR_TYPE_TMEM_INOUT;
+	params.u.tmem.buf_ptr = phpayload;
+	params.u.tmem.size = sizeof(struct tee_fs_rpc) + len;
+	params.u.tmem.shm_ref = cpayload;
+
+	/* fill in parameters */
+	*bf = *bf_cmd;
+
+	if (mode & TEE_FS_MODE_IN)
+		memcpy((void *)(bf + 1), data, len);
+
+	ret = thread_rpc_cmd(cmd_id, 1, &params);
+	/* update result */
+	*bf_cmd = *bf;
+	if (ret != TEE_SUCCESS)
+		goto exit;
+
+	if (mode & TEE_FS_MODE_OUT) {
+		uint32_t olen = MIN(len, bf->len);
+
+		memcpy(data, (void *)(bf + 1), olen);
+	}
+
+	res = TEE_SUCCESS;
+
+exit:
+	thread_rpc_free_payload(cpayload);
+	return res;
+}
+
+int tee_fs_rpc_access(int id, const char *name, int mode)
+{
+	struct tee_fs_rpc head = { 0 };
+	TEE_Result res;
+	int rc = RPC_FAILED;
+	size_t len;
+
+	DMSG("(id: %d, name: %s, mode: %d)...", id, name, mode);
```

Comment:
The dots are intended to better differentiate the function entry and exit. The parentheses are meant to group the parameters between the function name and the dots. I found this format rather easy to read as I was debugging this code. Here is an example log:
```
DEBUG:   [0x0] TEE-CORE:sql_fs_write:792: (fd: 0, buf: 0x3f07cfd8, len: 36)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_begin_transaction:138: (id: 8)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_begin_transaction:151: ...0
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:264: (id: 8, fd: 0, offset: 52, whence: 1)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:277: ...52
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_read:390: (id: 8, fd: 0, buf: 0x3f051430, len: 4124)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_read:411: ...0
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:264: (id: 8, fd: 0, offset: 52, whence: 1)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:277: ...52
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:495: (id: 8, fd: 0, buf: 0x3f051430, len: 4124)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:515: ...4124
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:264: (id: 8, fd: 0, offset: 0, whence: 1)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:277: ...0
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:495: (id: 8, fd: 0, buf: 0x3f053420, len: 52)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:515: ...52
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_end_transaction:184: (id: 8, rollback: 0)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_end_transaction:196: ...0
DEBUG:   [0x0] TEE-CORE:sql_fs_write:851: ...36
```